[DSCP-228] Make it possible to specify all node parameters in config file

flyingleafe opened the pull request:
What should be done:

The following executables should be able to launch with no CLI params but path to config file (and config key):

- [x] Witness node
- [x] Educator node
- [x] Faucet

For following executables overriding of separate config options via CLI params is possible:
- [x] Witness node
- [x] Educator node
- [x] Faucet

Also:
- [x] Config file structure in sane and well-documented
- [x] All launch scripts work in the same manner they did before

flyingleafe committed:
[DSCP-228] Comment regarding `Given` constraint with vinyl records

flyingleafe committed:
[DSCP-228] JSON instances for Witness resource params

Providing JSON instances for datatypes which are included
in `WitnessParams` is the least invasive way of getting rid of
`WitnessParams` and moving them to `WitnessConfig`, which doesn't
require rewriting all the resources code using `vinyl` configs.

flyingleafe committed:
[DSCP-228] Move `WitnessParams` entirely to `WitnessConfig`

flyingleafe committed:
[DSCP-228] JSON instances for Educator params

flyingleafe committed:
[DSCP-228] Move `EducatorParams` to `EducatorConfig`

Martoon-00 commented on core/package.yaml:
```diff
@@ -52,6 +52,7 @@ library:
     - time
     - unix
     - unliftio
+    - vinyl
```
Oh finally, now we can use `++` from vynil instead of self-written `+++`

Martoon-00 commented on core/src/Dscp/Config/Util.hs:
```diff
@@ -105,6 +110,15 @@ buildConfig ConfigParams{..} filler =
         pure config
 
+-- | Utility function for filling up a config reusing existing function
+-- for a subconfig.
```
I cannot parse this sentence :pape_sad:

Martoon-00 commented on core/src/Dscp/Util/Aeson.hs:
```diff
@@ -110,3 +111,9 @@ instance ToJSON SemVer.Version where
+    toJSON = String . toText . unitsF
+instance KnownRatName unit => FromJSON (Time unit) where
+    parseJSON = withText "time duration" $
+        nothingToFail ("Invalid time format" :: String) . unitsP . toString
```
Cool, didn't know about `unitsF` and `unitsP`

Martoon-00 commented on educator/src/Dscp/Educator/Config.hs:
```diff
@@ -1,30 +1,48 @@
+        , "keys" ::: EducatorKeyParams
+        , "api" ::: EducatorWebParams
+        ]
+     ]
```
Total configuration occurs to be ambiguous as soon as some parameters are present both in `witnessConfig` and `educatorConfig`.
Could we make `EducatorConfig` contain no fields of `WitnessConfig`; then parse `WitnessConfig ++ EducatorConfig` and `rcast` it appropriately when passing to `withWitnessConfig` and `withEducatorConfig`?

flyingleafe replied:
@Martoon-00 Well, I think you're wrong, because currently configuration doesn't contain any duplicated data (similar `api` and `keys` fields in `witness` and `educator` subsections are about different keys and different APIs). Though your suggestion is legit anyway; I even wanted to separate config further, having parts like `LoggingConfig`, `NetworkConfig`, etc.

Martoon-00 reviewed (commented):
LGTM, although I wonder how config would look like now.

flyingleafe committed:
[DSCP-228] Fix build after rebase

flyingleafe committed:
[DSCP-228] Pass configurations to `allocResource` explicitly

Implicit passing of configurations via reflections sometimes
leads to weird errors and/or ugly hacks when two `allocResource`
with different `Has*Config` constraints are nested.

Also, instance head parameter in `AllocResource` class has been
replaced by associated type family, because type family applications
cannot occur in instance head, restricting which types could be used
as params to `allocResource`.

flyingleafe committed:
[DSCP-228] Use type-level list concatenation from `vinyl`

flyingleafe committed:
[DSCP-228] Re-export `sub` and `option` in `Dscp.Config`

flyingleafe committed:
[DSCP-228] Remove unnecessary dependency on witness config

Functions in 'Dscp.Resource.Keys.Functions' depend only on core
config actually.

flyingleafe committed:
[DSCP-228] Move `FaucetParams` to `FaucetConfig`

flyingleafe committed:
[DSCP-228] Fix build after rebase

flyingleafe committed:
[DSCP-228] Backwards-compatible CLI parser for config

flyingleafe committed:
[DSCP-228] Backwards-compatible CLI args for Educator

flyingleafe committed:
[DSCP-228] Backwards-compatible CLI args for faucet

flyingleafe committed:
[DSCP-228] Sample config for witness node

flyingleafe committed:
[DSCP-228] Sample Educator config

flyingleafe committed:
[DSCP-228] Faucet config sample
